query/worker.go

```diff
@@ -161,6 +161,18 @@ func (w *worker) Run(results chan<- *jobResult, quit <-chan struct{}) {
 			// response handler to check whether it was answering
 			// our request.
 			case resp := <-msgChan:
+				select {
+				// If the peer disconnects before giving us a valid
+				// answer, we'll also exit with an error.
+				case <-peer.OnDisconnect():
+					log.Debugf("Peer %v for worker disconnected, "+
+						"cancelling job %v", peer.Addr(),
+						job.Index())
+
+					jobErr = ErrPeerDisconnected
+					break Loop
```

Comment:
i dont think this adds anything. This select is already covered in below and so if peer.OnDisconnect() triggers before the message is received on the msgChan, we will already catch the case below.

I think what you instead want is for `job.HandleResp`'s function to know if it can exit or not. 

The idiomatic way of doing this would be to pass it a `ctx context.Context` which gets cancelled when the peer disconnects. An intermediate step can just be to pass it a `quit chan struct{}` which you can derive from the `peer`s `OnDisconnect` method and then listen on that in the HandleResp call-back

Reply:
So the reason for adding this is because I saw that there are cases where `case resp := <- msgChan:` gets triggered first before the below `case <-peer.OnDisconnect():` gets triggered. 

> I think what you instead want is for job.HandleResp's function to know if it can exit or not.

Yeah that would also solve the problem I was seeing.

> An intermediate step can just be to pass it a quit chan struct{} which you can derive from the peers OnDisconnect method

That would be ok as well except now we force every implementation of `HandleResp` to handle the case where `quit chan struct{}` is passed.

Reply:
Well, I'll also add that I don't really have a preference on how that's handled so I'll push changes so that `HandleResp` takes a `quti chan struct{}` as an argument.